File: packages/kernel/src/logger.ts
import fs from "fs";
import path from "path";
import os from "os";

const SERVICE_LOG_BASE = path.join(os.homedir(), ".crona", "logs", "kernel");

function date(): string {
  return new Date().toISOString().slice(0, 10); // YYYY-MM-DD
}

function logDir(): string {
  const dir = path.join(SERVICE_LOG_BASE, date());
  try { fs.mkdirSync(dir, { recursive: true, mode: 0o700 }); } catch { /* ignore */ }
  return dir;
}

function formatEntry(level: string, msg: string, err?: unknown): string {
  const ts = new Date().toISOString();
  let entry = `[${ts}] [${level}] ${msg}`;
  if (err instanceof Error) {
    entry += `\n  Error: ${err.message}`;
    if (err.stack) entry += `\n  Stack: ${err.stack.split("\n").slice(1).join("\n  ")}`;
  } else if (err !== undefined) {
    entry += `\n  Detail: ${String(err)}`;
  }
  return entry + "\n";
}

export function logError(msg: string, err?: unknown): void {
  const entry = formatEntry("ERROR", msg, err);
  const dir = logDir();
  try { fs.appendFileSync(path.join(dir, "info.log"), entry); } catch { /* never throw from logger */ }
  try { fs.appendFileSync(path.join(dir, "error.log"), entry); } catch { /* never throw from logger */ }
}

export function logInfo(msg: string): void {
  const dir = logDir();
  try { fs.appendFileSync(path.join(dir, "info.log"), formatEntry("INFO", msg)); } catch { /* never throw from logger */ }
}

File: tui/internal/logger/logger.go
package logger

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"time"

	"crona/shared/config"
)

var logBase string

func init() {
	base, err := config.RuntimeBaseDir()
	if err != nil {
		base = "."
		if home, homeErr := os.UserHomeDir(); homeErr == nil {
			if resolved, resolvedErr := config.DefaultRuntimeBaseDirForModeOnOS(config.Load().Mode, runtime.GOOS, home, os.Getenv("XDG_DATA_HOME"), os.Getenv("LocalAppData")); resolvedErr == nil {
				base = resolved
			}
		}
	}
	logBase = filepath.Join(base, "logs", "tui")
}

func logDir() string {
	date := time.Now().UTC().Format("2006-01-02")
	dir := filepath.Join(logBase, date)
	_ = os.MkdirAll(dir, 0o700)
	return dir
}

func write(level, msg string) {
	ts := time.Now().UTC().Format(time.RFC3339Nano)
	entry := fmt.Sprintf("[%s] [%s] %s\n", ts, level, msg)
	dir := logDir()

	if f, err := os.OpenFile(filepath.Join(dir, "info.log"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600); err == nil {
		_, _ = f.WriteString(entry)
		_ = f.Close()
	}
	if level == "ERROR" {
		if f, err := os.OpenFile(filepath.Join(dir, "error.log"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600); err == nil {
			_, _ = f.WriteString(entry)
			_ = f.Close()
		}
	}
}

func Info(msg string)                   { write("INFO", msg) }
func Infof(format string, args ...any)  { write("INFO", fmt.Sprintf(format, args...)) }
func Error(msg string)                  { write("ERROR", msg) }
func Errorf(format string, args ...any) { write("ERROR", fmt.Sprintf(format, args...)) }
